src/mysql_vm_helpers.py

```diff
@@ -316,6 +421,44 @@ def _execute_commands(
         except subprocess.CalledProcessError as e:
             raise MySQLExecError(e.stderr)
 
+    def is_mysqld_running(self) -> bool:
+        """Returns whether mysqld is running."""
+        return os.path.exists(MYSQLD_SOCK_FILE)
+
+    def is_server_connectable(self) -> bool:
+        """Returns whether the server is connectable."""
+        # Always true since the charm runs on the same server as mysqld
+        return True
+
+    def stop_mysqld(self) -> None:
+        """Stops the mysqld process."""
+        logger.info(
+            f"Stopping service snap={CHARMED_MYSQL_SNAP_NAME}, service={CHARMED_MYSQLD_SERVICE}"
+        )
+
+        try:
+            snap_service_operation(CHARMED_MYSQL_SNAP_NAME, CHARMED_MYSQLD_SERVICE, "stop")
```

Comment:
is this sync, i.e. does it wait for the service be stopped or any further check is needed?

Reply:
i believe stopping a service is synchronous (at least i see messages that the command waits for the daemon to stop):

```
ubuntu@juju-173531-3:~$ ls -la /var/snap/charmed-mysql/common/mysql/
total 48
drwxr-xr-x 5 snap_daemon root 4096 Mar 15 19:34 .
drwxr-xr-x 3 root        root 4096 Mar 15 19:38 ..
drwxr-xr-x 2 snap_daemon root 4096 Mar 15 19:34 conf.d
drwxr-xr-x 8 snap_daemon root 4096 Mar 15 19:34 data
-rw-r----- 1 snap_daemon root 6632 Mar 15 19:34 error.log
-rw-r--r-- 1 snap_daemon root 1448 Mar 15 19:34 my.cnf.fallback
-rw-r--r-- 1 snap_daemon root 1610 Mar 15 19:34 mysql.cnf
drwxr-xr-x 2 snap_daemon root 4096 Mar 15 19:34 mysql.conf.d
-rw-r----- 1 snap_daemon root    5 Mar 15 19:34 mysqld.pid
srwxrwxrwx 1 snap_daemon root    0 Mar 15 19:34 mysqld.sock
-rw------- 1 snap_daemon root    5 Mar 15 19:34 mysqld.sock.lock
srwxrwxrwx 1 snap_daemon root    0 Mar 15 19:34 mysqlx.sock
-rw------- 1 snap_daemon root    5 Mar 15 19:34 mysqlx.sock.lock
ubuntu@juju-173531-3:~$ sudo snap stop charmed-mysql.mysqld 
2023-03-15T19:39:52Z INFO Waiting for "snap.charmed-mysql.mysqld.service" to stop.
Run service command "stop" for services ["mysqld"] of snap "charmed-mysql"                                                                                                                                \
Run service command "stop" for services ["mysqld"] of snap "charmed-mysql"                                                                                                                                \
Stopped.
ubuntu@juju-173531-3:~$ ls -la /var/snap/charmed-mysql/common/mysql/
total 36
drwxr-xr-x 5 snap_daemon root 4096 Mar 15 19:39 .
drwxr-xr-x 3 root        root 4096 Mar 15 19:38 ..
drwxr-xr-x 2 snap_daemon root 4096 Mar 15 19:34 conf.d
drwxr-xr-x 8 snap_daemon root 4096 Mar 15 19:39 data
-rw-r----- 1 snap_daemon root 7352 Mar 15 19:39 error.log
-rw-r--r-- 1 snap_daemon root 1448 Mar 15 19:34 my.cnf.fallback
-rw-r--r-- 1 snap_daemon root 1610 Mar 15 19:34 mysql.cnf
drwxr-xr-x 2 snap_daemon root 4096 Mar 15 19:34 mysql.conf.d

```